### ISSUE TITLE
Incorrect NuGet version detection (V2 vs V3)

### ISSUE THREAD

**Issue by ReedCopsey** [fsprojects/Paket#3842]
### Description

We are using a private NuGet source repo (feedz.io) which does not include /v3 in the URL for the source. When using Paket, it detects it as a V2 repo even though it should be V3.

This appears to be due to the logic here: https://github.com/fsprojects/Paket/blob/master/src/Paket.Core/Versioning/PackageSources.fs#L148-L151

Note that the code only assumes v3 if the URL has /v3 or is GitHub **and** ends in index.json. In this case, Feedz.io uses the index.json (which that URL includes the version info) file, but does not use /v3 in their API end points.

This also appears to be the same issue hit by other Paket users, whom have attempted to address (see https://github.com/fsprojects/Paket/pull/3806 and https://github.com/fsprojects/Paket/issues/3575).

### Repro steps

Repository at https://github.com/ReedCopsey/paket_bug demonstrates this. Using `dotnet paket install` gives many errors.

### Expected behavior

Paket would resolve this as V3 (somehow) and function properly.

### Actual behavior

Significant errors.

### Known workarounds

None.

**Comment by ReedCopsey** [comment]
Potential options to fix:

1. Assume V3 if it ends in index.json, such as #3806
2.  It does seem like pulling from the endpoint, and looking at the results would allow the version to be detected. For example, visiting https://f.feedz.io/ctech/test/nuget/index.json does include `{"version":"3.0.0",` at the start. This does appear to be part of the V3(+) spec as listed at https://github.com/NuGet/NuGetGallery/wiki/API-v3-Specification . 
3. Default to V3 instead of V2, given that it is likely to become more common over time
4. Provide a way to specify the version in the source (I'm fine saying "this is v3" in my sources)

**Comment by forki** [comment]
I think option 1 would be the least breaking.

**Comment by tebeco** [comment]
i'm not sure doing "url detect" is a good thing to be honest
```
https://foo-v3-bar/v2/index.json
https://foo-v3-bar/v2/
https://foo-v2-bar/v3/index.json
https://foobar/buzz
```

they could all be v2 or v3 feed, it's just a webserver replying to requests
Artifactory reply XML on url containing both `v3` and `index.json` for example ...

an idea would be to allow this :
`source https://foo/bar protocolVersion:3` 

`protocolVersion` specified ?
* use it
* not compliant => just throw "that source does not seems to be v3"

`protocolVersion` not specified ?
* attempt detection, but not too much
* failfast / throw / indicate that they have to specify `protocolVersion` on that source

**Comment by forki** [comment]
I'm also fine with the explicit switch.
Anyone up for implementing this?

TeBeCo <notifications@github.com> schrieb am Di., 5. Mai 2020, 22:19:

**Comment by ReedCopsey** [comment]
Personally, as a user, I'd expect V3 (or reasonably newest protocol) to be the default, with a switch to allow specification of others. Making V2 the default, especially given that it's (mostly) deprecated, seems odd.

**Comment by forki** [comment]
We may default to v3 protocol in paket v6 alpha. That would probably make a lot of sense

**Comment by tebeco** [comment]
@forki 
I tried to do something similar on https://github.com/fsprojects/Paket/pull/3844
Though I would love your inside on a PR on that PR here : https://github.com/tebeco/Paket/pull/2
Especially these points (in fact I wanted to remove all kind of "smart guess on URL" because that's the source of bug today) :
https://github.com/tebeco/Paket/pull/2#discussion_r421271567
https://github.com/tebeco/Paket/pull/2#discussion_r421330301
https://github.com/tebeco/Paket/pull/2#discussion_r421361695

**Comment by tebeco** [comment]
also i would love tripple check on the commit where the comment starts with `NEED HELP/REVIEW` :D as said ... I almost never used F#

**Comment by tebeco** [comment]
hello,

Would love feedback on the current state of the PR before continuing dev on it as I'm not sure how far it should go
Also I would love to know if you want ASAP the same PR on `master`  (to get it on v6) but with `Nuget protocolVersion 3` as default, and removing the notion of `known source` or not

**Comment by forki** [comment]
Sorry for Corona delay.

**Comment by forki** [comment]
Yes ideally we have 2 PRs. One I can apply on Bugfix branch with v2 as default. I would then release it and merge to master. Then a second one for master which changes the default

**Comment by tebeco** [comment]
> Sorry for Corona delay.

I hope you're safe and your familly too. That's what matter here
branch on `bugfixes` is ready for review

I'll start "revamping" the "v6" one

**Comment by nyctef** [comment]
I'm not sure (I'm new to this whole discussion) but I think we might have just been broken by #3844 / 5.247.0 - our previously-working v3 source seems to now be treated as a v2 source: (example line from build log)
`RequestFailedException: Request to 'https://red-gate.pkgs.visualstudio.com/_packaging/Main/nuget/v3/index.json/FindPackagesById()?semVerLevel=2.0.0&id='RedGate.Installer'' failed with: 404 NotFound -`

Does this mean we now need to change all our `paket.dependencies` files to manually specify the protocol version?

**Comment by forki** [comment]
Ok I'll revert

**Comment by forki** [comment]
I released 5.247.1 

@tebeco we need to find a way to make it as non breaking as possible. Or we can only do it for v6.

**Comment by nyctef** [comment]
Thanks for the quick response ❤️ Sorry for the trouble

**Comment by tebeco** [comment]
As said in the "protocolVersion" PR that other previous PR should not have been merged
because it's now a breaking change

that the big issue in paket v5:
Relying on "magic url guess"

That protocolVersion is supposed to be used when the Url is not "magically detected"

This visual studio feed url is "not in the magic list"
just adding "protocolVersion: 3" at the end of the line would have simply work ;)

see comment here:
https://github.com/fsprojects/Paket/pull/3844#issuecomment-633530118

**Comment by forki** [comment]
Yes but we don't want to require people to do that. In v5 we would need to
add it to the magic list. It's just really bad if we break builds in minor
update

TeBeCo <notifications@github.com> schrieb am So., 14. Juni 2020, 13:41:

**Comment by tebeco** [comment]
FYI:
I'll recreate / open the exact same PR that was rollback and add this Url to the Test to see what's going on

Also i'll check with adding explicit "protocolVersion: 3"

**Comment by tebeco** [comment]
I'll try to fix that Url
not sure why it was not originally in the unit test to be fair if it was a known "magic one"

do you have idea about other one that could have been forgotten ?
